nntrainer/layers/rnn.cpp

```diff
@@ -163,11 +164,68 @@ void RNNLayer::copy(std::shared_ptr<Layer> l) {
 }
 
 void RNNLayer::calcDerivative() {
-  //  NYI
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &weight =
+    weightAt(static_cast<int>(RNNParams::weight_xh)).getVariableRef();
+  Tensor &ret_ = net_input[0]->getGradientRef();
+
+  ret_ = derivative_.dot(weight, ret_, false, true);
 }
 
 void RNNLayer::calcGradient() {
-  // NYI
+  Tensor &djdw_x =
+    weightAt(static_cast<int>(RNNParams::weight_xh)).getGradientRef();
+  Tensor &djdw_h =
+    weightAt(static_cast<int>(RNNParams::weight_hh)).getGradientRef();
+  Tensor &djdb_h =
+    weightAt(static_cast<int>(RNNParams::bias_h)).getGradientRef();
+  Tensor &weight_hh =
+    weightAt(static_cast<int>(RNNParams::weight_hh)).getVariableRef();
+
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+  Tensor dh_nx = Tensor(TensorDim(1, 1, 1, derivative_.width()));
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+    Tensor deriv_t = derivative_.getBatchSlice(b, 1);
+    Tensor xs_t = input_.getBatchSlice(b, 1);
+    Tensor hs_t = hidden_.getBatchSlice(b, 1);
+    dh_nx.setZero();
+
+    Tensor dh;
+    Tensor xs;
+    Tensor hs_prev;
+    Tensor hs;
+
+    for (unsigned int t = deriv_t.height(); t-- > 0;) {
```

Comment:
```suggestion
    for (int t = deriv_t.height(); t-- > 0;) {
```
It seems like t might go underflow

Reply:
I don't think so. t never be negative.

Reply:
Right, I mistook.